### ISSUE TITLE
How to build Warlock Spells?

### ISSUE THREAD

**Issue by Merott** [ngbp/warlock#1]
Hey Josh,

This looks very promising, and I'm quite keen to start testing and using Warlock for a little AngularJS pet project.

I've been using `ngbp` build system for some time, and found it quite powerful, but then I became interested in Gulp and [started creating a build system](https://github.com/Cyberlaze/nGulp) on top of it, mostly following the `ngbp` app structure for the application part.

I just started using my Gulp-based build system in a small project, but looking at Warlock, I think I might just ditch that build system of mine, and start using Warlock, but I'd need to know how to build my own Spells, so that I can automate the most common tasks that I'd need for my project. I've looked at your [coffeescript spell](https://github.com/ngbp/spell-coffeescript), and it doesn't seem very complicated.

It would be really nice if you could set up a guide together on how to create Warlock Spells. I'd then jump right in, and of course I'd be more than happy to share my Spells!

**Comment by joshdmiller** [comment]
Hey Merott,

I very much appreciate you taking the time to check things out - and for submitting pull requests!

Essentially all documentation is missing, partially because the API is in such flux. As I add new spells, I discover I need new ways of handling something in Warlock internally, and so I go and change the API. In the early stages, this works reasonably well, but once it's out in the wild, it's a disaster. Another reason to keep the group of testers reasonably small.

You're right, though, that spells aren't complicated. In fact, they're really straightforward once the concepts are understood. I just added some basic documentation to the wiki; check it out and let me know what you think. There's still a _lot_ more documentation to write.

**Comment by Merott** [comment]
I like Warlock's simple use. It has some rather complicated internals, but I guess that complexity comes naturally when you build something that is intended to _"just work"_. So I'm happy to dig in and understand it, while doing anything I can to help improve it.

I started creating a spell yesterday with whatever I knew from digging around Warlock and its spells' code, but didn't quite get there until I read your guides, so thanks for that!

I wanted to create a spell that takes in `*.stub.json` or `*.stub.js` files, and processes them through a Lodash template to generate stubbed AngularJS modules. I use this to stub my AngularJS repository/resources layer, so that I can develop backend-less. The code generation part goes into the template file, so it has nothing to do with Warlock or the build system. All Warlock needs to do is to get the stub definition files and process them with the template, then push them into the `scripts-to-build` flow after jslint is complete, so that they'll be output to the build directory.

At the moment I'm still a little fuzzy around step priorities and flow merge priorities (i.e. `merge: 'flow::scripts-to-build::20'` vs `.add( 80, 'angular-stubs-generate', tplWrap )`), but I guess that'll slowly sink in :smile:

In any case I managed to _develop my first spell_, so I paste my code below. Feel free to comment on anything I could have done differently. 

### `stubs.coffee`

``` coffee
wrap = require 'gulp-wrap'

modules_dir = (require 'path').resolve 'node_modules'
defaultTplPath = "#{modules_dir}/warlock-spell-angular-stubs/plugins/stub.js.tpl";

module.exports = ( warlock ) ->
  tplWrap = (options) ->
    options.src = options.tpl
    options.src = defaultTplPath if options.src is "default"
    delete options.tpl
    wrap(options)

  warlock.flow 'angular-stubs-to-generate',
    source: [ '<%= globs.source.stubs %>' ]
    source_options:
      base: "<%= paths.source_app %>"
    tasks: [ 'webapp-build' ]
    merge: 'flow::scripts-to-build::20'

  .add( 80, 'angular-stubs-generate', tplWrap )
```

### `package.json`

``` json
{
  "name": "warlock-spell-angular-stubs",
  "version": "1.0.0-alpha.1",
  "description": "An addon to add Angular module stubbing support to the Warlock Angular spell",
  "homepage": "https://github.com/merott/spell-angular-stubs",
  "author": {
    "name": "Merott Movahedi",
    "url": "http://github.com/merott"
  },
  "repository": {
    "type": "git",
    "url": "git://github.com/merott/spell-angular-stubs.git"
  },
  "bugs": {
    "url": "https://github.com/merott/spell-angular-stubs"
  },
  "licenses": [
    {
      "type": "MIT",
      "url": "https://github.com/merott/spell-angular-stubs/blob/master/LICENSE"
    }
  ],
  "engines": {
    "node": ">= 0.10.0"
  },
  "peerDependencies": {
    "warlock-spell-angular": "~1.0.0"
  },
  "keywords": [
    "warlockplugin",
    "warlockspell"
  ],
  "warlock": {
    "globs": {
      "source": {
        "js": [
          "!<%= paths.source_app %>/**/*.stub.js"
        ],
        "stubs": [
          "<%= paths.source_app %>/**/*.stub.js",
          "<%= paths.source_app %>/**/*.stub.json"
        ]
      }
    },
    "tasks": {
      "angular-stubs-generate": {
         "tpl": "default"
      }
    }
  },
  "dependencies": {
    "gulp-wrap": "~0.3.0"
  }
}
```

P.S. I'm new to CoffeeScript too, so I had to learn a little CS to understand some of your code. I haven't decided if it's worth learning or not, but I love Ruby, so I'm supposed to like the CoffeeScript syntax. Yet, I'm undecided.

**Comment by joshdmiller** [comment]
While you're messing around with spells, you can use local plugins. If you put your code in, say, `plugins/stubs.coffee` or whatever in your local project, you can tell warlock to load it by adding the directory to the `plugins` array of your `warlock.json`. E.g.:

``` js
{
  "plugins": [ "plugins" ]
}
```

Then you don't have to use NPM. Obviously, all your configuration and dependencies would need to go in your project's `package.json`, so it's not for reusable components. For one-off stuff, it can be pretty handy, though.

As for CoffeeScript, I'm not a fanboy. I like the succinct syntax, but it's not for everyone and I go back and forth. Honestly, I don't remember why I chose to do everything in CoffeeScript. haha

But anyway, as a matter of form, I would place the `stubs.js.tpl` in a folder _other_ than `plugins` to keep it clean. The path can then be:

``` coffee
defaultTplPath = warlock.file.joinPath __dirname, "../templates/stub.js.tpl"
```

As for flow and merge priorities, don't overthink them. If you create a flow, each time you use `add` you are adding a stream/task to the queue; these are sorted by priority before they are run. So:

``` coffee
warlock.flow( "myFlow" )
.add( 5, "blah1", something1 )
.add( 20, "blah3", something3 ) # out of order
.add( 15, "blah2", something2 )
```

Will, in essence, do this:

``` js
files.pipe( something1() ).pipe( something2() ).pipe( something3() )
```

This allows the user to, in their files, add tasks to existing flows. For merges, we're asking the flow _into which_ we're merging to create a task at the specified priority and, at that point, to take whatever was already in that stream add to it whatever this stream ended up with. For example, if I had a task that merged into `flow::myFlow::17`, then `myFlow`'s `something1` and `something2` would _only_ get its own files, whereas `something3` would get its own files _and_ whatever files were merged:

``` js
files.pipe( something1() ).pipe( something2() ).addOtherFlow().pipe( something3() )
```

By the way, if you can think of ways to improve the docs or if there are areas you find unclear, please let me know.

**Comment by Merott** [comment]
The local plugins feature is handy, and good to know, I was pointing NPM directly to my local working directory.

I found your guides helpful, obviously helpful enough to discover my own inner powers and develop a spell of my own. I sure feel like a Warlock now. I'll definitely let you know if there is anything unclear, I still need to write a few more spells for my project. Talking of _unclear_, I think you got the order of your pipes mixed up in your example above :smiley: 

I get your point though. What I was confusing was the relation between the priority numbers defined when adding tasks to a `flow`, and the priority number you enter when you specify a `merge`. It makes total sense though, I don't even know what I was confused about now. Well, that's confusing :confused: 

Thank you for your elaborations, and as I said, I'll let you know if I have any ideas, or come across anything in your docs that's not clear.

**Comment by joshdmiller** [comment]
Crap! It was actually the priority that I got wrong for the point I was trying to illustrate. I fixed it. Well, that's what I get for aiming for clarity.

I misunderstood your initial question, indeed. The add and merge priorities are the same, except the merge priority is like an add for the _other_ flow, rather than the current one. But it sounds like you already figured all that out, so I'll shut up now. :-)

Speaking of shutting up, since I threw up some docs, I can close this issue. 1 down - woot!
